File: Behave-python/features/steps/login_logout_steps.py
from behave import given, when, then, fixture, use_fixture
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC

LOGIN_URL = "https://www.saucedemo.com/"
USERNAME = "standard_user"
PASSWORD = "secret_sauce"

@fixture
def browser_chrome(context):
    options = webdriver.ChromeOptions()
    options.add_argument('--headless')
    context.driver = webdriver.Chrome(options=options)
    context.driver.implicitly_wait(5)
    yield context.driver
    context.driver.quit()

def before_all(context):
    use_fixture(browser_chrome, context)

def after_all(context):
    if hasattr(context, 'driver'):
        context.driver.quit()

@given('I am on the login page')
def step_impl(context):
    context.driver.get(LOGIN_URL)
    WebDriverWait(context.driver, 10).until(
        EC.presence_of_element_located((By.ID, "user-name"))
    )

@when('I login with username "{username}" and password "{password}"')
def step_impl(context, username, password):
    context.driver.find_element(By.ID, "user-name").send_keys(username)
    context.driver.find_element(By.ID, "password").send_keys(password)
    context.driver.find_element(By.ID, "login-button").click()

@then('I should see the dashboard')
def step_impl(context):
    WebDriverWait(context.driver, 10).until(
        EC.presence_of_element_located((By.CLASS_NAME, "app_logo"))
    )
    assert "inventory" in context.driver.current_url

@given('I am logged in as "{username}" with password "{password}"')
def step_impl(context, username, password):
    context.driver.get(LOGIN_URL)
    WebDriverWait(context.driver, 10).until(
        EC.presence_of_element_located((By.ID, "user-name"))
    )
    context.driver.find_element(By.ID, "user-name").send_keys(username)
    context.driver.find_element(By.ID, "password").send_keys(password)
    context.driver.find_element(By.ID, "login-button").click()
    WebDriverWait(context.driver, 10).until(
        EC.presence_of_element_located((By.CLASS_NAME, "app_logo"))
    )

@when('I logout')
def step_impl(context):
    menu_btn = WebDriverWait(context.driver, 10).until(
        EC.element_to_be_clickable((By.ID, "react-burger-menu-btn"))
    )
    menu_btn.click()
    logout_btn = WebDriverWait(context.driver, 10).until(
        EC.element_to_be_clickable((By.ID, "logout_sidebar_link"))
    )
    logout_btn.click()

@then('I should be redirected to the login page')
def step_impl(context):
    WebDriverWait(context.driver, 10).until(
        EC.presence_of_element_located((By.ID, "login-button"))
    )
    assert "saucedemo.com" in context.driver.current_url

File: Cucumber-Java/src/test/java/steps/LoginLogoutSteps.java
package steps;

import io.cucumber.java.AfterAll;
import io.cucumber.java.BeforeAll;
import io.cucumber.java.en.*;
import org.openqa.selenium.*;
import org.openqa.selenium.support.ui.*;
import util.WebDriverFactory;

public class LoginLogoutSteps {
    private static WebDriver driver;
    private static WebDriverWait wait;
    private static final String BASE_URL = "https://www.saucedemo.com/";

    @BeforeAll
    public static void setUp() {
        driver = WebDriverFactory.create();
        wait = new WebDriverWait(driver, 10);
    }

    @AfterAll
    public static void tearDown() {
        if (driver != null) driver.quit();
    }

    @Given("I am on the login page")
    public void i_am_on_login_page() {
        driver.get(BASE_URL);
        wait.until(ExpectedConditions.visibilityOfElementLocated(By.id("user-name")));
    }

    @When("I login with username {string} and password {string}")
    public void i_login_with_credentials(String username, String password) {
        driver.findElement(By.id("user-name")).sendKeys(username);
        driver.findElement(By.id("password")).sendKeys(password);
        driver.findElement(By.id("login-button")).click();
    }

    @Then("I should see the dashboard")
    public void i_should_see_dashboard() {
        wait.until(ExpectedConditions.visibilityOfElementLocated(By.className("app_logo")));
        assert driver.getCurrentUrl().contains("inventory");
    }

    @Given("I am logged in as {string} with password {string}")
    public void i_am_logged_in(String username, String password) {
        i_am_on_login_page();
        i_login_with_credentials(username, password);
        i_should_see_dashboard();
    }

    @When("I logout")
    public void i_logout() {
        WebElement menuBtn = wait.until(ExpectedConditions.elementToBeClickable(By.id("react-burger-menu-btn")));
        menuBtn.click();
        WebElement logoutBtn = wait.until(ExpectedConditions.elementToBeClickable(By.id("logout_sidebar_link")));
        logoutBtn.click();
    }

    @Then("I should be redirected to the login page")
    public void i_should_be_redirected_to_login() {
        wait.until(ExpectedConditions.visibilityOfElementLocated(By.id("login-button")));
        assert driver.getCurrentUrl().contains("saucedemo.com");
    }
}
